docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,844 @@
+
+After all the providers are in paused state, you can proceed with the upgrade
+as described in the previous paragraph (change the `spec.Version` field).
+
+Once all the providers are upgraded to a version that abides to the new
+contract, it is possible for the operator to unpause providers; the operator
+does not allow to unpause providers if there are still providers abiding to
+the old contract.
+
+Please note that we are planning to embed this sequence (pause - upgrade -
+unpause) as a part of `clusterctl upgrade apply` command when there is a
+contract change.
+
+
+##### Changing a provider
+
+On top of changing a provider version (upgrades), the operator supports also
+changing other provider fields, most notably controller flags and variables.
+
+The operation internally works like upgrades: The current instance of the
+provider is deleted, while preserving CRDs, namespaced and user objects A new
+instance of the provider is installed with the new set of flags/variables.
+
+Please note that clusterctl currently does not support this operation.
+
+See Example 1 in [Example API Usage](#example-api-usage)
+
+##### Deleting a provider
+
+In order to delete a provider you have to delete the corresponding provider
+object.
+
+Deletion of the provider will be blocked if any workload cluster using the
+provider still exists.
+
+Additionally, deletion of a core provider should be blocked if there are still
+other providers in the management cluster.
+
+#### Operator Lifecycle Management
+
+##### Operator Installation
+
+- `clusterctl init` will install the operator and its corresponding CRDs as a
+  pre-requisite if the operator doesn’t already exist. Please note that this
+  command will consider image overrides defined in the local clusterctl config
+  file.
+- If the admin does not want to use clusterctl to install the operator, it is
+  possible to kubectl apply the operator yaml that will be published in the
+  cluster-api release artifacts.
+- There can be only one operator installed per management cluster, across all
+  namespaces.
+
+##### Operator Upgrade
+
+- The admin can use `clusterctl upgrade operator` to upgrade the operator
+  components. Please note that this command will consider image overrides
+  defined in the local clusterctl config file.
+- If the admin doesn’t want to use clusterctl, they can use kubectl apply with
+  the latest version of the operator yaml that will be published in the
+  cluster-api release artifacts.
+- clusterctl will require a matching operator version. In the future, when
+  clusterctl move to beta/GA, we will reconsider supporting version skew
+  between clusterctl and the operator.
+
+##### Operator Delete
+
+- clusterctl will delete the operator as part of the `clusterctl delete --all`
+  command.
+- If the admin doesn’t want to use clusterctl, they can use kubectl delete.
+  However, it’s the admin’s responsibility to verify that there are no
+  providers running in the management cluster.
```

Comment:
What if the admin tries to use clusterctl CLI after the operator is deleted? should we document this here?

Reply:
I'm assuming that if the user is using the latest clusterctl (_one which uses the operator_) after the operator is deleted, the clusterctl will respond with an error saying that it is unable to detect the operator and that the operator needs to be installed.
Unless of course, the user does a `clusterctl init`, after which it will install the operator again.

Reply:
makes sense